sig-scheduler-plugins/pkg/controllers/podgroup_controller.go

```diff
@@ -0,0 +1,473 @@
+}
+
+// +kubebuilder:rbac:groups=scheduling.x-k8s.io,resources=podgroups,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=scheduling.x-k8s.io,resources=podgroups/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=scheduling.x-k8s.io,resources=podgroups/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// Note that we currently don't do deletion based on owner references, but that
+// would be ideal (I could not get it to work)
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+func (r *PodGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.Info("reconciling flux-framework/fluence-controller for request")
+	pg := &schedv1alpha1.PodGroup{}
+
+	// Get the timestamp as soon as reconcile happens as a fallback below
+	timestamp := metav1.NewMicroTime(time.Now())
+	if err := r.Get(ctx, req.NamespacedName, pg); err != nil {
+
+		// Case 1: if we get here and it's not found, assume not created
+		if apierrs.IsNotFound(err) {
+			log.Info("PodGroup", "Status", fmt.Sprintf("Pod group %s is not found, deleted.", req.NamespacedName))
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, fmt.Sprintf("Unable to retrieve pod group %s", req.NamespacedName))
+		return ctrl.Result{}, err
+	}
+	log.Info("REFERENCES", "Reconciler", pg.ObjectMeta.OwnerReferences)
+
+	// Grab all statuses (and groups of them) we are interested in
+	schedulingOrPending := (pg.Status.Phase == schedv1alpha1.PodGroupScheduling || pg.Status.Phase == schedv1alpha1.PodGroupPending)
+	twoDaysOld := pg.Status.ScheduleStartTime.Sub(pg.CreationTimestamp.Time) > 48*time.Hour
+	finishedOrFailed := pg.Status.Phase == schedv1alpha1.PodGroupFinished || pg.Status.Phase == schedv1alpha1.PodGroupFailed
+
+	// Finished or failed - clean up the group
+	if finishedOrFailed {
+		log.Info("PodGroup", "Status", fmt.Sprintf("Pod group %s is finished or failed.", req.NamespacedName))
+		return ctrl.Result{}, nil
+	}
+
+	// If startScheduleTime - createTime > 2days,
+	// do not reconcile again because pod may have been GCed
+	if schedulingOrPending && pg.Status.Running == 0 && twoDaysOld {
+		r.recorder.Event(pg, v1.EventTypeWarning, "Timeout", "schedule time longer than 48 hours")
+		return ctrl.Result{}, nil
+	}
+
+	// We can get the podList and check for sizes here
+	podList := &v1.PodList{}
+
+	// Select based on the group name
+	groupNameSelector := labels.Set(map[string]string{schedv1alpha1.PodGroupLabel: pg.Name}).AsSelector()
+	err := r.List(ctx, podList, client.MatchingLabelsSelector{Selector: groupNameSelector})
+	if err != nil {
+		log.Error(err, "List pods for group failed")
+		return ctrl.Result{}, err
+	}
+
+	// If the scheduler time created is Zero (not set) we set it here
```

Comment:
This is a really good question - and I'm really glad you asked, because this was a design decision that I made (and didn't properly articulate) and now I'm not sure it is the best choice. Let me walk through the logic, and my thinking at the time.

If you look at [the upstream](https://github.com/kubernetes-sigs/scheduler-plugins/blob/2a7767e4b85b80ba86d967d860855b05899bb488/pkg/controllers/podgroup_controller_test.go#L330) this particular field isn't populated beyond a test. The only place it is used in the original original controller is here:

https://github.com/kubernetes-sigs/scheduler-plugins/blob/2a7767e4b85b80ba86d967d860855b05899bb488/pkg/controllers/podgroup_controller.go#L83-L89

It's to say "after this timestamp, if we've gone 48 hours over it, that's when we can call it garbage collected and clean up." Since I was trying to make minimal changes to the fields, I hijacked that field to _instead_ be the timestamp of when the podGroup is first reconciled, which (I think) is a pretty good estimate for when the podGroup is created because it has to reconcile. The field is owned by the controller, so it will always be null / unset "Zero" on the first reconcile, so we only check to ensure it happens on the first reconcile. The reason I used this field is because the CreationTime field used by the analogous coscheduling plugin ([here](https://github.com/kubernetes-sigs/scheduler-plugins/blob/2a7767e4b85b80ba86d967d860855b05899bb488/pkg/coscheduling/core/core.go#L245-L256)) is a field on the object that I can't mutate (at least I think CreationTimestamp of an object is pretty standard for Kubernetes objects) and I needed to mutate it to change the granularity from one second. I could either add a new field or use an existing one (that was not being used) and I chose the latter.

So it was essentially a choice between making minimal changes (adding a field) vs. using a (what I deemed likely unused) status field to [return here](https://github.com/flux-framework/flux-k8s/blob/2aee785257ff4617adf3b4445d096cf36403adaf/sig-scheduler-plugins/pkg/fluence/core/core.go#L344) instead of the CreationTimestamp, which I could not change. At the time I didn't see the need for the scheduleStartTime, but now that we talked about it in a previous comment, I can see it could be useful to customize. I'm leaning toward now that we want to keep this field as is, maintain the current logic to check that against 48 hours, and then add a new field for our CreationTime. Let me know your thoughts!

Reply:
I'll push a commit with my suggested fix - just still working through the rest of the review.

Reply:
> I'm leaning toward now that we want to keep this field as is, maintain the current logic to check that against 48 hours, and then add a new field for our CreationTime.

Yes, I like that idea. I think it will minimize breakage if the unused field starts getting used in the future.